Add Service Creation performance test

alculquicondor opened the pull request:
Fixes #2617

## Proposed Changes

* Add tests that create services and report the average duration (from prometheus) to testgrid. Tests for 1 and 20 services.
* Update CreateLatestServiceWithResources to use the image name from ResourceNames.

/assign @srinivashegde86 
/cc @mdemirhan

srinivashegde86 commented:
I already have a PR out for this using the perf framework https://github.com/knative/serving/pull/2810. We can add more stuff to the output, like the prom query you have here.

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+
+	deployGrp, _ := errgroup.WithContext(context.Background())
+
+	logger.Infof("Creating %v services", noServices)
```
%d, perhaps?

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	if err := deployGrp.Wait(); err != nil {
+		t.Fatalf("Error waiting for services to become ready")
```
no formatting, t.Fatal()

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+
+	promAPI, err := prometheus.PromAPI()
+	if err != nil {
+		logger.Error("Cannot setup prometheus API")
```
log error?

alculquicondor replied:
Fatal instead.

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+	query := fmt.Sprintf("sum(controller_%s{key=~\"%s/%s-.*\"})", reconciler.ServiceReadyLatencyN, test.ServingNamespace, basename)
+	logger.Infof("Query: %v", query)
+	val, err := prometheus.RunQuery(context.Background(), logger, promAPI, query)
+	if err != nil {
```
fail the test?
Also it's more readable, imo:
```
if err != nil {
  ...
  return
}
// Process no error case.
```

alculquicondor replied:
Fatal

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+
+const (
+	tCreateServiceN    = "TestCreateService"
+	tCreate20ServicesN = "TestCreate20Services"
```
perhaps, tCreateNServicesN = "TestCreate%dServices" and then below fmt.Sprintf().
Since if tomorrow we decide to change 20 to 25, it'll be more places to change, perhaps missing this one?

alculquicondor replied:
Done

vagababov commented on test/performance/service_creation_test.go:
```diff
@@ -0,0 +1,106 @@
+	}
+}
+
+func TestCreateService(t *testing.T) {
```
Perhaps just `TestCreateServices`, and inside
```
	for _, s := range []int{1, 20} {
		n := fmt.Sprintf(tCreateNServicesN, s)
		t.Run(n, func(t *testing.T) {
			testCreateServices(t, n, s)
		})
	}
```
?
Also at this point, since the test will be named with the size, we can omit passing the name, since it's only used for logging.

alculquicondor replied:
The name is also used for testgrid. Keeping it.

alculquicondor commented:
/hold @srinivashegde86 and I are debating what is the best approach.

alculquicondor commented:
/hold

alculquicondor commented:
@mdemirhan This PR is a good opportunity to also surface reconciler metrics such as the number of reconciles and their latency. What do you think? We could be collecting them for the individual created services or for the test.

However, @mattmoor raised the concern of sending unbounded keys to prometheus (like the name of the service) in term of storage and querying performance. How would you suggest to isolate the metrics otherwise?

alculquicondor commented:
For reference, this is how I'm performing the query for the whole test https://github.com/knative/serving/pull/2852/files#diff-acba5ba347f259f106a265df17caa24aR86
I agree that it might lead to performance issues. Maybe querying by a time window would be better, but I couldn't find how.

mdemirhan commented:
I am not sure if I understand the concern. The data is already getting sent to prometheus and storage cost is already being accounted for. Looking at your query `sum(controller_%s{key=~\"%s/%s-.*\"})", reconciler.ServiceReadyLatencyN, test.ServingNamespace, basename`, it seems like you are querying data for only a specific service and such a query should be quick enough. You should probably measure the cost of the query - that should give us some data points and we can discuss further based on the results of that.

Once concern I have is regarding the usability of this data: when you do a sum like that, you will get a single data point - is good enough information? If you deploy and delete the same service name for example, you will get an aggregate single data point which might not be useful.

alculquicondor commented:
It's not a specific service. It's for the set of services created during a single run of the test (note the `=~`). Prometheus's docs don't mention anything about the performance of a query like that. It guess it wouldn't be much of a concern if knative artifacts are deployed and teared down every time for the test suit.

I don't think there's any issue with the usability. This metric is specific to service readiness. But maybe I'm missing something? What happens when a service is upgraded? Does it become not-ready?

mdemirhan commented:
I am not worried about the size of the query for this - it looks like a small enough query.

As for the usability of this metric, the issue with a single aggregate is that you will get a single number representing how long it took for service to be ready. What if this service is deployed twice - the number is going to be misleading (will be a sum of both). More information such as count, total latency, 90% latency (if the metric type is histogram) would be better.

alculquicondor commented:
Note that we are only looking into a single test, where we control which services are deployed. `sum` was just a quick way to get an average for all the services created during the test. It can be replaced by `avg`, which I didn't see before.

As for the 90% latency; would that require a change of aggregation here? https://github.com/knative/serving/blob/master/pkg/reconciler/stats_reporter.go#L75
My interest, though, is to get aggregated metrics for all the services created during the test.

mdemirhan commented:
I didn't realize that the stat was a gauge. In that case, it will show you the value of the last latency and you can ignore my comments. You shouldn't do avg or anything else, but just show the last value you get from prometheus.

alculquicondor commented:
/hold cancel

@mdemirhan Each of the services has a last value, but we can still aggregate for all the services created during a single test. I changed sum for avg, and added max and quantile 0.95.

@mattmoor I'm setting the tests for 1 and 10 services. Let me know if those numbers are reasonable (maybe too low?).

srinivashegde86 commented:
/cc @mattmoor

mattmoor commented:
We should merge this with https://github.com/knative/serving/pull/2957, which shares the same core across both tests (and that logic has changed a bunch in the past few days).

I'm running 10, 50, 100 right now, but I'd like to see us running at least 150 and 200 sooner than later.

alculquicondor commented:
Keeping this PR on hold as it might be of use in the future, for 2 reasons:
- It can leverage prometheus' aggregation functions (like percentiles) to avoid ad-hoc implementations in the tests themselves.
- It can provide sub-second resolution without polling the API.

Feel free to close if these are not compelling reasons.

srinivashegde86 commented:
/test pull-knative-serving-perf-tests

markusthoemmes commented:
What do we want to do with this PR? @srinivashegde86 any idea?

srinivashegde86 commented:
we have a perf scale test that tests #services that can be created. But, that does a bunch of manual calculations and does not use prom metrics, while this one does. I would like to defer to the WG here and we can close this if they feel like the current test is sufficient and we dont want this additional detail.

@mattmoor do we need this or can we close this?

markusthoemmes commented:
poke, it's been another month of inactivity.
